### PR TITLE
[feat] Benchmarking BlockSparse Kernels for Sparse Transformers

### PR DESCRIPTION
**Pull request by anijain2305**
## What does this PR do?
Benchmark Torch, Triton and Sputnik kernels for Sparse Transformers.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

cc @cpuhrsch

### CODE REVIEW

**Comment by blefaudeux** [comment]
Great @anijain2305, thanks for the PR ! It's just missing [some lint fixes](https://app.circleci.com/pipelines/github/facebookresearch/xformers/1367/workflows/0c7c7c2f-5df5-4eb4-91d9-7dcf5943e8de/jobs/2993/parallel-runs/0/steps/0-111), I would recommend pre-commit if you can set this up, it saved some of my sanity :) We're migrating to triton2, blocksparse should be a little faster if I remember @ptillet correctly and definitely more stable, and these tools will definitely be useful !

File: xformers/benchmarks/benchmark_blocksparse_transformers.py [hunk 1]
```diff
@@ -0,0 +1,1042 @@
...
+
+device = "cuda"
+TestCase = namedtuple("TestCase", ["prepare_callable", "mask", "config", "name"])
+
+##############################################
+# Plotting utilities
+##############################################
+
+
+def pretty_barplot(results, title, units: str, filename=None, dash_key=""):
```
Comment (blefaudeux) [inline xformers/benchmarks/benchmark_blocksparse_transformers.py:31]: nit but we had some utils (not as nice but these could be a complement) [here](https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/utils.py). Maybe better to move this there for easier code discovery ?

File: xformers/benchmarks/benchmark_blocksparse_transformers.py [hunk 2]
```diff
@@ -0,0 +1,1042 @@
...
+        seq_length = self.config.seq_length
+        if keep_blocked:
+            seq_length = self.config.blocked_seq_length
+        return self.expand(causal_1d_pattern(seq_length))
+
+    def __str__(self):
+        return "lower_triangular"
+
+
+class BigBirdAttentionMask(AttentionMask):
```
Comment (blefaudeux) [inline xformers/benchmarks/benchmark_blocksparse_transformers.py:255]: @ptillet an example of blocksparse + mask, it's being used by real people I promise :)

**Review by blefaudeux — APPROVED** [review]
LGTM, if you're ok to relocate the generic plots to benchmarks/utils and fix the linting ? Thanks again, this is great work

File: xformers/benchmarks/benchmark_blocksparse_transformers.py (hunk 2 above)
Comment (ptillet) [reply xformers/benchmarks/benchmark_blocksparse_transformers.py:255]: :upside_down_face: is there any way to define this mask statically?

Comment (blefaudeux) [reply xformers/benchmarks/benchmark_blocksparse_transformers.py:255]: yes actually, I think that it's a very acceptable limitation. In vision there are some approaches to make the attention very dynamic which could benefit from floating masks (handwavy explanation, like RPN for attention), but I think that not supporting that is decent enough.. [Butterfly matrices ](https://dawn.cs.stanford.edu/2019/06/13/butterfly/)is another application but it can be static as far as I remember

Comment (ptillet) [reply xformers/benchmarks/benchmark_blocksparse_transformers.py:255]: Yep. Just to be clear, supporting completely dynamic masks is possible, but the overhead is high so I want to incentivize using static attention masks as much as possible. If there is an analytical formula to determine which indices should be omitted, this can be done pretty easily as shown there https://github.com/openai/triton/blob/master/python/triton/ops/blocksparse/softmax.py#L67

**Comment by blefaudeux** [comment]
hey @anijain2305 let me know if you would prefer some help with the linting / minor cleanup !

**Comment by anijain2305** [comment]
Oh, so sorry. This completely slipped my mind.

I will get this done by Monday for sure (out for a couple of days).

**Comment by blefaudeux** [comment]
> Oh, so sorry. This completely slipped my mind.
> 
> 
> 
> I will get this done by Monday for sure (out for a couple of days).

No worries, it was just in case ! Thanks, looking forward!

**Comment by blefaudeux** [comment]
The doc build issue is unrelated, it's a new issue from CI which I never had, fixing it on another branch. Sorry about that, getting there !
